### PR TITLE
fix(platform-server): destroy `PlatformRef` when error happens during the `bootstrap()` phase, e.g. in `APP_INIIALIZER` - to prevent memory leaks in SSR

### PR DESCRIPTION
**Pull request by Platonn**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [N/A] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

`PlatformRef` is not destroyed if an error happens during the `bootstrap()` phase (e.g. in `APP_INITIALIZER`). Then the main app's injector is not destroyed and therefore `ngOnDestroy` hooks of singleton services were not called on the end (failure) of SSR. 

Note: This could lead to possible memory leaks in custom SSR apps, if their singleton services' `ngOnDestroy` hooks contained an important teardown logic (e.g. unsubscribing from RxJS observable).

Issue Number: https://github.com/angular/angular/issues/58111

## What is the new behavior?
- `PlatformRef` is guaranteed to be destroyed in `renderApplication()` and `renderModule()` functions of `@angular/platform-server` both in the case of successful render and in the case of a failure (resulting in a rejected promise)
- `PLATFORM_DESTROY_LISTENERS` that destroys the app's main injector is now set not only in case of `renderApplication()`, but also in case of `renderModule()`

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

**Comment by JeanMeche** [comment]
Hi Krzysztof, thanks for looking into this. Could you please rebase your PR on the `main` branch (so we can run the CI)

**Comment by Platonn** [comment]
Thank you @JeanMeche for your reply.
I've rebased with `main` just now and resolved the code conflicts.

File: packages/platform-server/test/integration_spec.ts [hunk 1]
```diff
@@ -1076,6 +1079,153 @@ class HiddenModule {}
...
+              imports: [MyServerAppModule, ServerModule],
+              bootstrap: [MyServerApp],
+            })
+            class ServerFailingAppInitializerModule {}
+
+            const ServerFailingAppInitializerAppStandalone = getStandaloneBootstrapFn(
+              createMyServerApp(true),
+              FailingAppInitializerProviders,
+            );
+            debugger;
```
Comment (JeanMeche) [inline packages/platform-server/test/integration_spec.ts:1167]: You forgot this one

Comment (Platonn) [reply packages/platform-server/test/integration_spec.ts]: 🤦‍♂️ Thank you for noticing it. Now removed.

**Comment by JeanMeche** [comment]
The changes are breaking the `elements` pacakge, can you have a look ?

**Comment by Platonn** [comment]
Thank you @JeanMeche for letting me know.
Now pushed a fix.

Explanation:
- Before this PR
  - the `isApplicationBootstrapConfig()` was duck-typing based on the presence of the property `platformInjector`.
- In this PR initially
  -  I've changed the logic to check the property `rootComponent` instead, because I wanted to pass the `platformInjector` in all cases - also in the NgModule flow (in order to setup the `PLATFORM_DESTROY_LISTENER` later in the `bootstrap()` function).
   - It turned out to break the behavior of the public API function `createApplication()` which was tested inside the `elements` package. The test failed, because no `rootComponent` was passed in that test case setup as a param of `createApplication`. Then `isApplicationBootstrapConfig()` returned `false` which let into executing NgModule-flow of `bootstrap()`. Despite someone could perceive omitting the `rootComponent` as an edge case usage of `createApplication()`, I broke that default behavior. So needed to fix it
- The fix
  - So now I've fixed the problem by no-longer duck-typing inside the `isApplicationBootstrapConfig()` function. Instead, I started passing explicitly a newly-introduced config property `BootstrapKind` as a required param of the `bootstrap()` function. btw. It's not a breaking change, because the signature of the `bootstrap()` function is not a part of the public API, as far as I can understand.

File: packages/core/src/platform/bootstrap.ts [hunk 2]
```diff
@@ -26,21 +26,32 @@ import {Injector} from '../di';
 import {InternalNgModuleRef, NgModuleRef} from '../linker/ng_module_factory';
 import {stringify} from '../util/stringify';
 
-export interface ModuleBootstrapConfig<M> {
+export enum BootstrapKind {
```
Comment (JeanMeche) [inline packages/core/src/platform/bootstrap.ts:29]: nit:
fwiw, if the enum is not exported outside of the package, we can make it a `const enum` with a numeric value. (It'll save a few bytes)

Comment (Platonn) [reply packages/core/src/platform/bootstrap.ts]: Thank you for a hint.
Pushed the fix now.

File: packages/platform-server/src/utils.ts [hunk 3]
```diff
@@ -210,18 +210,20 @@ async function _render(platformRef: PlatformRef, applicationRef: ApplicationRef)
   }
 
   appendServerContextInfo(applicationRef);
-  const output = platformState.renderToString();
+  return platformState.renderToString();
```
Comment (alan-agius4) [inline packages/platform-server/src/utils.ts:213]: NIT

```suggestion
  
  return platformState.renderToString();
```

File: packages/platform-server/src/utils.ts [hunk 4]
```diff
@@ -210,18 +210,20 @@ async function _render(platformRef: PlatformRef, applicationRef: ApplicationRef)
...
+}
 
-  // Destroy the application in a macrotask, this allows pending promises to be settled and errors
-  // to be surfaced to the users.
+/**
+ * Destroy the application in a macrotask, this allows pending promises to be settled and errors
+ * to be surfaced to the users.
+ */
+async function asyncDestroyPlatform(platformRef: PlatformRef): Promise<void> {
   await new Promise<void>((resolve) => {
```
Comment (alan-agius4) [inline packages/platform-server/src/utils.ts:221]: Avoid creating an extra microtask.

```suggestion
function asyncDestroyPlatform(platformRef: PlatformRef): Promise<void> {
  return new Promise<void>((resolve) => {
```

**Review by alan-agius4 — APPROVED** [review]
LGTM, just a couple of NITs

File: packages/core/src/platform/bootstrap.ts [hunk 5]
```diff
@@ -26,21 +26,32 @@ import {Injector} from '../di';
...
   r3Injector: R3Injector;
-  platformInjector: Injector;
   rootComponent: Type<unknown> | undefined;
 }
 
 function isApplicationBootstrapConfig(
   config: ApplicationBootstrapConfig | ModuleBootstrapConfig<unknown>,
 ): config is ApplicationBootstrapConfig {
-  return !!(config as ApplicationBootstrapConfig).platformInjector;
+  return config.bootstrapKind === BootstrapKind.Application;
```
[the remark this answers is not in the document]
Comment (Platonn) [reply packages/core/src/platform/bootstrap.ts]: Thank you. done

File: packages/platform-server/src/utils.ts (hunk 4 above)
Comment (Platonn) [reply packages/platform-server/src/utils.ts]: Thank you. done

**Comment by AndrewKushnir** [comment]
@Platonn thanks for creating this PR 👍 It looks like there is a failing test (see [the `test` CI job output](https://github.com/angular/angular/actions/runs/11237605771/job/31241076382?pr=58112)). Could you please take a look at it when you get a chance?

**Comment by JeanMeche** [comment]
@AndrewKushnir It looks like the failing test is flaky.

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:683881745:BASE:683881771:1728448838691:abd9cd7b).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: presubmit looks good (only pre-existing failing targets), this PR is ready for merge.

**Comment by devversion** [comment]
This PR was merged into the repository by commit 9e82559de4e99a1aedf645a05b01fc08d3f4b1b1.

The changes were merged into the following branches: main

**Comment by devversion** [comment]
@Platonn @JeanMeche this seems to have conflicts with the patch branch, so I only merged for v19. Please send another PR if necessary— I tried creating a patch port but I don't know how conflicts should be solved

**Comment by JeanMeche** [comment]
@devversion Patch #58135

**Pull request closed without merging**